idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddExclExclCallFixFactories.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix.fixes
+
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.fir.api.fixes.diagnosticFixFactory
+import org.jetbrains.kotlin.idea.frontend.api.KtAnalysisSession
+import org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.KtFirDiagnostic
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtFunctionSymbol
+import org.jetbrains.kotlin.idea.frontend.api.types.KtType
+import org.jetbrains.kotlin.idea.quickfix.AddExclExclCallFix
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+object AddExclExclCallFixFactories {
+    private fun PsiElement?.asFix(hasImplicitReceiver: Boolean = false) =
+        this?.let { listOf(AddExclExclCallFix(it, hasImplicitReceiver)) } ?: emptyList()
+
+    val unsafeCallFactory = diagnosticFixFactory<KtFirDiagnostic.UnsafeCall> { diagnostic ->
+        getFixForUnsafeCall(diagnostic.psi)
+    }
+
+    val unsafeInfixCallFactory = diagnosticFixFactory<KtFirDiagnostic.UnsafeInfixCall> { diagnostic ->
+        getFixForUnsafeCall(diagnostic.psi)
+    }
+
+    val unsafeOperatorCallFactory = diagnosticFixFactory<KtFirDiagnostic.UnsafeOperatorCall> { diagnostic ->
+        getFixForUnsafeCall(diagnostic.psi)
+    }
+
+    private fun KtAnalysisSession.getFixForUnsafeCall(psi: PsiElement): List<IntentionAction> {
+        val (target, hasImplicitReceiver) = when (psi) {
+            // `foo.bar` -> `foo!!.bar`
+            is KtDotQualifiedExpression -> psi.receiverExpression to false
+
+            // `foo[bar]` -> `foo!![bar]`
+            is KtArrayAccessExpression -> psi.arrayExpression to false
+
+            is KtCallableReferenceExpression -> psi.lhs.let { lhs ->
+                if (lhs != null) {
+                    // `foo::bar` -> `foo!!::bar`
+                    lhs to false
+                } else {
+                    // `::bar -> this!!::bar`
+                    psi to true
+                }
+            }
+
+            // `bar` -> `this!!.bar`
+            is KtNameReferenceExpression -> psi to true
```

Comment:
Maybe it's not in the scope of this PR either. But the quickfix does not work for the following

```
interface A {
    var s: String
}

fun foo(a: A?, s: String?) {
    a.s = s
}
```

The root cause seems to be that for `FirVariableAssignment` there is no source corresponding to the lValue PSI element. Hence, the error is reported on the named reference. Ideally, it should have been reported on the qualified name instead.

Reply:
Well there should be two errors here:
1. UNSAFE_CALL in `a.s`
2. TYPE_MISMATCH on the assignment

Are you referring to (2)?  I noted this in a few of the test files that FIR doesn't report TYPE_MISMATCH yet on assignments.

Reply:
Sorry I am referring to 1. The example code is not good. I should have declared `s: String` in `foo`.

Reply:
Ohhh ok you were referring to (1). I see it now that the error is on `s` and not on `a.s`, that the `lValue` isn't the entire left-hand expression as you mentioned. I'll look into it but most likely this should be done on a separate PR. I will add a test though.

Reply:
Alright, I was able to fix this issue. We need to report the error on the entire LHS expression of the assignment. For that added `lhsExpression` to `FirVariableAssignment` so we can use it for reporting.